### PR TITLE
[SYCL] Enqueue leaves after host accessor destruction

### PR DESCRIPTION
**Pull request by KarachunIvan**
Need to enqueue all leaves of the graph to start
execution after unblocking.

Signed-off-by: Ivan Karachun <ivan.karachun@intel.com>

### CODE REVIEW

**Comment by romanovvlad** [comment]
LGTM. Please fix title: leaves-> leafs

**Comment by romanovvlad** [comment]
Could you please write a test which check that PI call which enqueues kernel is called during host accessor d'tor?

**Comment by keryell** [comment]
> LGTM. Please fix title: leaves-> leafs

It sounds like it is mostly *leaves* https://www.merriam-webster.com/dictionary/leaf

Otherwise, I think that the Intel FPGA team is in Toronto...
https://www.quora.com/Why-are-the-Toronto-Maple-Leafs-spelled-as-Leafs-instead-of-Leaves
:-)

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -141,6 +141,19 @@ EventImplPtr Scheduler::addHostAccessor(Requirement *Req) {
...
 void Scheduler::releaseHostAccessor(Requirement *Req) {
   Req->MBlockedCmd->MCanEnqueue = true;
+  MemObjRecord* Record = Req->MSYCLMemObj->MRecord.get();
+  for (Command *Cmd : Record->MReadLeafs) {
+    EnqueueResultT Res;
+    bool Enqueued = GraphProcessor::enqueueCommand(Cmd, Res);
+    if (!Enqueued && EnqueueResultT::FAILED == Res.MResult)
+      throw runtime_error("Enqueue process failed.");
+  }
+  for (Command *Cmd : Record->MWriteLeafs) {
```
Comment (keryell) [inline sycl/source/detail/scheduler/scheduler.cpp]: `Record->MWriteLeaves` ?

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 2]
```diff
@@ -141,6 +141,19 @@ EventImplPtr Scheduler::addHostAccessor(Requirement *Req) {
...
+    EnqueueResultT Res;
+    bool Enqueued = GraphProcessor::enqueueCommand(Cmd, Res);
+    if (!Enqueued && EnqueueResultT::FAILED == Res.MResult)
+      throw runtime_error("Enqueue process failed.");
+  }
```
Comment (keryell) [inline sycl/source/detail/scheduler/scheduler.cpp]: You could factorize the code in a lambda and call it on `Record->MReadLeafs` and `Record->MWriteLeafs`

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 3]
```diff
@@ -141,6 +141,19 @@ EventImplPtr Scheduler::addHostAccessor(Requirement *Req) {
 
 void Scheduler::releaseHostAccessor(Requirement *Req) {
   Req->MBlockedCmd->MCanEnqueue = true;
+  MemObjRecord* Record = Req->MSYCLMemObj->MRecord.get();
+  for (Command *Cmd : Record->MReadLeafs) {
```
Comment (keryell) [inline sycl/source/detail/scheduler/scheduler.cpp]: `Record->MReadLeaves` ?

Comment (KarachunIvan) [reply sycl/source/detail/scheduler/scheduler.cpp]: You are right, but to fix that I have to change the name of the variable in whole library.

**Comment by romanovvlad** [comment]
> 
> 
> > LGTM. Please fix title: leaves-> leafs
> 
> It sounds like it is mostly _leaves_ https://www.merriam-webster.com/dictionary/leaf
> 
> Otherwise, I think that the Intel FPGA team is in Toronto...
> https://www.quora.com/Why-are-the-Toronto-Maple-Leafs-spelled-as-Leafs-instead-of-Leaves
> :-)

Yeah, you are right.

File: sycl/source/detail/scheduler/scheduler.cpp (hunk 3 above)
Comment (romanovvlad) [reply sycl/source/detail/scheduler/scheduler.cpp]: Yes, please separate refactoring from functional changes.

File: sycl/test/scheduler/HostAccDestruction.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: env SYCL_PI_TRACE=1 %CPU_RUN_PLACEHOLDER %t.out 2>&1 %CPU_CHECK_PLACEHOLDER
```
Comment (romanovvlad) [inline sycl/test/scheduler/HostAccDestruction.cpp]: I think running on cpu device should be enough. Suggest removing other run lines.

File: sycl/source/detail/scheduler/scheduler.cpp (hunk 3 above)
Comment (KarachunIvan) [reply sycl/source/detail/scheduler/scheduler.cpp]: Sure thing, will do with another patch.

File: sycl/test/scheduler/HostAccDestruction.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t.out
```
Comment (romanovvlad) [inline sycl/test/scheduler/HostAccDestruction.cpp]: It seems run on host is not needed.

Comment (KarachunIvan) [reply sycl/test/scheduler/HostAccDestruction.cpp]: Removed

**Pull request merged**